rest/diagnostic_api.go

```diff
@@ -0,0 +1,347 @@
+//  Copyright 2013-Present Couchbase, Inc.
+//
+//  Use of this software is governed by the Business Source License included
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/channels"
+	"github.com/gorilla/mux"
+)
+
+// each field in this response is a map of collection names to a map of channel names to grant history info
+type getAllChannelsResponse struct {
+	AdminGrants       map[string]map[string]auth.GrantHistory            `json:"admin_grants,omitempty"`
+	DynamicGrants     map[string]map[string]auth.GrantHistory            `json:"dynamic_grants,omitempty"`
+	JWTGrants         map[string]map[string]auth.GrantHistory            `json:"jwt_grants,omitempty"`
+	AdminRoleGrants   map[string]map[string]map[string]auth.GrantHistory `json:"admin_role_grants,omitempty"`
+	DynamicRoleGrants map[string]map[string]map[string]auth.GrantHistory `json:"dynamic_role_grants,omitempty"`
+}
+
+func (h *handler) handleGetAllChannels() error {
+	h.assertAdminOnly()
+	user, err := h.db.Authenticator(h.ctx()).GetUser(internalUserName(mux.Vars(h.rq)["name"]))
+	if err != nil {
+		return err
+	}
+	if user == nil {
+		return kNotFoundError
+	}
+	var resp getAllChannelsResponse
+	resp.AdminGrants = make(map[string]map[string]auth.GrantHistory, len(user.ExplicitChannels()))
+	resp.DynamicGrants = make(map[string]map[string]auth.GrantHistory, len(user.Channels())-len(user.ExplicitChannels()))
+	resp.DynamicRoleGrants = make(map[string]map[string]map[string]auth.GrantHistory, len(user.RoleNames())-len(user.ExplicitRoles()))
+	resp.AdminRoleGrants = make(map[string]map[string]map[string]auth.GrantHistory, len(user.ExplicitRoles()))
+	authenticator := h.db.Authenticator(h.ctx())
+	userPrinc, err := authenticator.GetPrincipal(user.Name(), true)
+	if err != nil {
+		return err
+	}
+
+	if h.db.OnlyDefaultCollection() {
+		defaultKeyspace := "_default._default"
+		if len(user.ExplicitChannels()) > 0 {
+			resp.AdminGrants[defaultKeyspace] = make(map[string]auth.GrantHistory)
+		}
+		resp.DynamicGrants[defaultKeyspace] = make(map[string]auth.GrantHistory)
+		err = authenticator.RebuildCollectionChannels(userPrinc, base.DefaultScope, base.DefaultCollection)
+		if err != nil {
+			return err
```

Comment:
The way this looks like it can fail is if the query fails, in this case we might want to know what call hit this. It's 

```suggestion
			return fmt.Errorf("Could not rebuild channels for %s: %w", keyspace, err)
```

It looks hard to trigger a failure here but it is worth adding an error message if this does fail.